src/backend/functions/orchestrator/lib/index.js

```diff
@@ -102,35 +98,103 @@ function createDeleteParams({mediaUrl, tasksTableName}) {
    };
 }
 
-// terminated :: ({k:v} -> Promise {k: v}) -> {k: v} -> String -> Promise {k:v}
+// terminated :: ({k:v} -> Promise {k: v}) -> {k: v} -> {k: v} -> Promise {k:v}
 const terminated = R.curry(deleteItem =>
    handle(o(deleteItem, createDeleteParams), ({MediaUrl: mediaUrl}) => ({mediaUrl}))
 );
 
-function error(ecs, ddb) {}
+const errorUpdate = R.curry((updateItem, {tasksTableName, mediaUrl, taskArn}) => {
+   const params = {
+      TableName: tasksTableName,
+      Key: {
+         MediaUrl: mediaUrl
+      },
+      UpdateExpression: 'ADD Retries :val SET TaskStatus = :status, TaskArn = :task',
+      ExpressionAttributeValues: {
+         ':val': 1,
+         ':status':  'WAITING',
+         ':task': taskArn
+      },
+      ReturnValues: 'ALL_NEW'
+   };
+
+   return updateItem(params)
+       .then(({Attributes}) => ({...Attributes, tasksTableName}))
+});
 
+const unrecoverableError = R.curry((updateItem, {tasksTableName, MediaUrl}) => {
+   const params = {
+      TableName: tasksTableName,
+      Key: {
+         MediaUrl: MediaUrl
+      },
+      UpdateExpression: 'SET TaskStatus = :status',
+      ExpressionAttributeValues: {
+         ':status':  'UNRECOVERABLE_ERROR'
+      },
+      ReturnValues: 'ALL_NEW'
+   };
+
+   return updateItem(params);
+});
+
+// error :: AWS.ECS -> AWS.DynamoDB -> {k: v} -> {k: v} -> Promise {k:v}
+const error = R.curry((ecs, updateItem) =>
+    handle(composeP(errorUpdate(updateItem), startTranscription(ecs)), ({MediaUrl: mediaUrl}) => ({mediaUrl}))
+);
 
 // handler :: AWS.ECS -> AWS.DynamoDB -> {k: v} -> ({k: v} -> {k: v} -> Promise {k: v})
 module.exports = (ecs, ddb, env) => {
+   const update = params => ddb.update(params).promise();
+
    const handlers = {
-      WAITING: waiting(ecs, params => ddb.update(params).promise()),
+      WAITING: waiting(ecs, update),
       INITIALIZING: noop,
       PROCESSING: noop,
       TERMINATING: terminating(ecs),
       TERMINATED: terminated(params => ddb.delete(params).promise()),
-      ERROR: error
+      ERROR: error(ecs, update),
+      UNRECOVERABLE_ERROR: noop
    };
 
-   return (event, context) => {
+   return event => {
+
       const promises = R.map(record => {
          const image = converter.unmarshall(R.path(['dynamodb', 'NewImage'], record));
-         const handler = R.defaultTo(noop, handlers[image.TaskStatus]);
-         return handler(env, image);
+         const {MediaUrl, Retries, TaskStatus} = image;
+         const {TASKS_TABLE_NAME: tasksTableName} = env;
+
+         console.log('Status: ' + TaskStatus);
+
+         if(R.defaultTo(0, Retries) > parseInt(env.RETRY_THRESHOLD)) {
+            console.log('Error threshold exceeded');
+            return unrecoverableError(update, {MediaUrl, tasksTableName});
+         }
+
+         const handler = R.defaultTo(noop, handlers[TaskStatus]);
+         return handler(env, image).catch(err => {
```

Comment:
we should probably log the error here because if the catch block returns the error, the log message won't show up anywhere from my testing. returning the error will be treated as a success by Lambda so there'll be no automatic retries - I think that's ok though. thoughts?

Reply:
Yep, makes sense.